common/src/test/java/com/thoughtworks/go/domain/materials/git/GitCommandTest.java

```diff
@@ -25,777 +25,879 @@
 import org.junit.rules.TemporaryFolder;
 import org.junit.rules.TestRule;
 
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.charset.Charset;
 import java.util.*;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import static com.thoughtworks.go.domain.materials.git.GitTestRepo.*;
 import static com.thoughtworks.go.util.DateUtils.parseRFC822;
-import static com.thoughtworks.go.util.ReflectionUtil.getField;
 import static com.thoughtworks.go.util.command.ProcessOutputStreamConsumer.inMemoryConsumer;
+import static java.lang.String.format;
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.regex.Pattern.compile;
+import static java.util.regex.Pattern.quote;
 import static org.apache.commons.io.filefilter.FileFilterUtils.*;
 import static org.apache.commons.lang3.time.DateUtils.addDays;
 import static org.apache.commons.lang3.time.DateUtils.setMilliseconds;
-import static org.assertj.core.api.Assertions.*;
-import static org.mockito.MockitoAnnotations.initMocks;
+import static org.junit.jupiter.api.Assertions.*;
 
-@EnableRuleMigrationSupport
 public class GitCommandTest {
-    private static final String BRANCH = "foo";
-    private static final String SUBMODULE = "submodule-1";
-
-    private GitCommand git;
-    private String repoUrl;
-    private File repoLocation;
-    private static final Date THREE_DAYS_FROM_NOW = setMilliseconds(addDays(new Date(), 3), 0);
-    private GitTestRepo gitRepo;
-    private File gitLocalRepoDir;
-    private GitTestRepo gitFooBranchBundle;
-
-    @Rule
-    public final TemporaryFolder temporaryFolder = new TemporaryFolder();
-
-    @Rule
-    public final TestRule restoreSystemProperties = new RestoreSystemProperties();
-
-    @Rule
-    public final ExpectedException expectedException = ExpectedException.none();
-
-    @BeforeEach
-    void setup() throws Exception {
-        gitRepo = new GitTestRepo(temporaryFolder);
-        gitLocalRepoDir = createTempWorkingDirectory();
-        git = new GitCommand(null, gitLocalRepoDir, GitMaterialConfig.DEFAULT_BRANCH, false, null);
-        repoLocation = gitRepo.gitRepository();
-        repoUrl = gitRepo.projectRepositoryUrl();
-        InMemoryStreamConsumer outputStreamConsumer = inMemoryConsumer();
-        int returnCode = git.cloneWithNoCheckout(outputStreamConsumer, repoUrl);
-        if (returnCode > 0) {
-            fail(outputStreamConsumer.getAllOutput());
-        }
-        gitFooBranchBundle = GitTestRepo.testRepoAtBranch(GIT_FOO_BRANCH_BUNDLE, BRANCH, temporaryFolder);
-        initMocks(this);
-    }
+    private static GitCommand withBranch(String branch) {
+        return new GitCommand(null, null, branch, false, null);
+    }
+
+    private static GitCommand withBranch(String branch, File cwd) {
+        return new GitCommand(null, cwd, branch, false, null);
+    }
+
+    @Nested
+    class RefSpec {
```

Comment:
Apologies for the messy diff;

The main bits here are adding the `RefSpec` and `RefSpecIntegration` `@Nested` tests. The rest is reorganizing the original tests under the `@Nested class Integration {}` tests.